File: BOJ/BOJ15686.cpp
//#include<iostream>
//#include<vector>
//#include<algorithm>
//#include<limits.h>
//
//using namespace std;
//
//int M, min_dis = INT_MAX;
//int dis[100][13];
//vector <pair<int, int>> c, h;
//vector <int> sel;
//
//int cal() {
//	int dist = 0;
//	int sum = 0;
//	for (int i = 0; i < h.size(); i++) {
//		int minn = INT_MAX;
//		//M개의 치킨집을 모두 탐색
//		for (int j = 0; j < sel.size(); j++) {
//			minn = min(minn, dis[i][sel[j]]); //가장 작은 치킨 거리 저장;
//		}
//		sum = sum + minn; //집마다의 치킨거리를 더해줌(도시의 치킨 거리 계산)
//	}
//
//	return sum;
//}
//
//void solve(int idx) {
//	//M개를 뽑았으면
//	if (sel.size() == M) {
//		min_dis = min(min_dis, cal()); //가장 작은 도시의 치킨 거리 저장
//		return;
//	}
//
//	//치킨집 조합
//	for (int i = idx; i < c.size(); i++) {
//		sel.push_back(i); //벡터에 현재 치킨집 추가
//		solve(i+1); //재귀
//		sel.pop_back(); //조합 벡터 원상복귀
//	}
//}
//
//int main() {
//	ios_base::sync_with_stdio(false);
//	cin.tie(0);
//	cout.tie(0);
//
//	int N;
//
//	cin >> N >> M;
//
//	for (int i = 0; i < N; i++) {
//		for (int j = 0; j < N; j++) {
//			int num;
//			cin >> num;
//			if (num == 2) {
//				c.push_back(make_pair(i, j)); //모든 치킨집 저장
//			}
//			else if (num == 1) {
//				h.push_back(make_pair(i, j)); //모든 집 저장
//			}
//		}
//	}
//
//	//각 집마다 치킨 집과의 거리를 계산하고 저장
//	for (int i = 0; i < h.size(); i++) {
//		for (int j = 0; j < c.size(); j++) {
//			dis[i][j] = abs(h[i].first - c[j].first) + abs(h[i].second - c[j].second);
//		}
//	}
//
//	solve(0);
//
//	cout << min_dis  << "\n";
//
//	return 0;
//}

File: BOJ/BOJ15686.java
package com.ssafy.day01;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class BOJ15686 {
	static ArrayList<pair> p;
	static int N, M, minn, min_dis;
	static int[][] city;
	static boolean[][] visit;
	
	static class pair{
		int x;
		int y;
		
		public pair() {}
		public pair(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		city = new int [N][N];
		visit = new boolean[N][N];
		min_dis = Integer.MAX_VALUE;
		p = new ArrayList<pair>();
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				city[i][j]= Integer.parseInt(st.nextToken());
			}
		}
		
		solve(0,0);
		bw.write(min_dis + "\n");
		br.close();
		bw.flush();
		bw.close();
	}
	public static void solve(int idx, int cnt) {
		if (cnt == M) {
			min_dis = Math.min(min_dis, cal());
			return;
		}
		
		for (int i = idx; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (city[i][j] == 2 && visit[i][j] == false) {
					visit[i][j] = true;
					p.add(new pair(i,j));
					solve(i, cnt + 1);
					visit[i][j] = false;
					p.remove(p.size()-1);
				}
			}
		}
	}
	public static int cal() {
		int dis = 0;
		int sum = 0;
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < N; j++) {
				if (city[i][j] == 1) {
					int k = 0;
					minn = Integer.MAX_VALUE;
					while (k < M) {
						dis = Math.abs(i - p.get(k).x) + Math.abs(j - p.get(k).y);
						minn = Math.min(minn, dis);
						k++;
					}
					sum = sum + minn;
				}
			}
		}
		
		return sum;
	}

}
